orttraining/orttraining/python/ort_trainer.py

```diff
@@ -515,6 +517,70 @@ def create_ort_training_session_bind_parameters(model, device, world_rank=-1, wo
 
     return session, train_io_binding, eval_io_binding, output_name, torch_params, output_types
 
+def save_checkpoint(model, checkpoint_dir, checkpoint_prefix="ORT_checkpoint", checkpoint_state_dict=None):
```

Comment:
As per offline discussion, we would like to see the Checkpoint API decoupled from ORTTrainer. That makes it more portable in general, especially for the upcoming ORTModel, in which the user can pass the optimizer as part of checkpoint_state_dict

To accomplish this, we envision the following changes are necessary.

1) Merge model and checkpoint_state_dict into a single 'obj' as the first parameter of the save_checkpoint API
1.1) obj can be either ORTTrainer, ORTModel or an dictionary with them inside. For the ORTModel scenario, additionally to ORTModel instance itself, an instance of torch.optimi.Optimizer has to be present in the dictionary
1.2) The model and (for ORTModel scenario) optimizer can assume fixed (and documented) names as 'model' and 'opmizer' inside the dict or they can have arbitrary name that are discovered by iterating over the dict during input validation
1.3) Besides model and optimizer state, users can add arbitrary (key, value) entries on 'obj' for persistence
2) checkpoint_state_dict can be renamed to tag to be similar to DeepSpeed and without a default value. That helps supporting multiple checkpoints in the same folder through tag pattern matching
3) As ORTModel doesnt support ZeRO, it is OK to add a static variable called ORTModel.partition_optimizer=False to ensure that both ORTModel and ORTTrainer have mathcing APIs to support checkpoint. We understand that both models already suppoert world_rank and world_size